docs/source/function_overloading.rst

```diff
@@ -1,60 +1,63 @@
-Function overloading in stubs
-=============================
+Function Overloading
+====================
 
-Sometimes you have a library function that seems to call for two or
-more signatures.  That's okay -- you can define multiple *overloaded*
-instances of a function with the same name but different signatures in
-a stub file (this feature is not supported for user code, at least not
-yet) using the ``@overload`` decorator. For example, we can define an
-``abs`` function that works for both ``int`` and ``float`` arguments:
+Sometimes the types in a function depend on each other in ways that can't be 
+captured with a simple ``Union``.  For example, the ``__getitem__`` (``[]`` bracket 
+indexing) method can take an integer and return a single item, or take a ``slice`` 
+and return a ``Sequence`` of items.  You might be tempted to annotate it like so:
 
 .. code-block:: python
 
-   # This is a stub file!
+    class Seq(Generic[T], Sequence[T]):
+        def __getitem__(self, index: Union[int, slice]) -> Union[T, Sequence[T]]:
+            pass
+    
+But this is a little loose, as it implies that when you put in an ``int`` you might
+sometimes get out a single item or sometimes a sequence.  To capture a constraint 
+such as a return type that depends on a parameter type, we can use 
+`overloading <https://www.python.org/dev/peps/pep-0484/#function-method-overloading>`_
+to give the same function multiple type annotations (signatures).
 
-   from typing import overload
-
-   @overload
-   def abs(n: int) -> int: pass
-
-   @overload
-   def abs(n: float) -> float: pass
+.. code-block:: python
 
-Note that we can't use ``Union[int, float]`` as the argument type,
-since this wouldn't allow us to express that the return
-type depends on the argument type.
+    from typing import Generic, Sequence, overload
+    T = TypeVar('T')
 
-Now if we import ``abs`` as defined in the above library stub, we can
-write code like this, and the types are inferred correctly:
+    class Seq(Generic[T], Sequence[T]):
+        @overload       # These are just for the type checker, and overwritten by the real implementation
```

Comment:
Such long comments will look a bit ugly on ``readthedocs``. I would recommend wrapping them or shortening to approx. 80 chars.